File: src/bumperbot_py_examples/bumperbot_py_examples/simple_parameter.py
import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import SetParametersResult
from rclpy.parameter import Parameter
class SimpleParameter(Node):
    def __init__(self):
        super().__init__('simple_parameter')
        self.declare_parameter("simple_int_param", 28)
        self.declare_parameter("simple_string_param", "Anas")

        self.add_on_set_parameters_callback(self.paramChangeCallback)

    def paramChangeCallback(self, params):
        result = SetParametersResult()

        for param in params:
            if param.name == "simple_int_param" and param.type_ == Parameter.Type.INTEGER:
                self.get_logger().info(f"Old param: {self.get_parameter('simple_int_param').get_parameter_value().integer_value}, New param: {param.value}")
                result.successful = True

            if param.name == "simple_string_param" and param.type_ == Parameter.Type.STRING:
                self.get_logger().info(f"New param: {param.value}")
                result.successful = True
        return result


def main(args=None):
    rclpy.init(args=args)
    node = SimpleParameter()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()
    return node

if __name__ == '__main__':
    main()

File: src/bumperbot_cpp_examples/src/simple_parameter.cpp
#include <string>
#include <vector>
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include <rcl_interfaces/msg/set_parameters_result.hpp>


using std::placeholders::_1;

class SimpleParameter : public rclcpp::Node {
public:
    SimpleParameter() : Node("simple_parameter") {
        declare_parameter<int>("simple_int_param", 42);
        declare_parameter<std::string>("simple_string_param", "Ali");

        param_callback_handle_ = add_on_set_parameters_callback(std::bind(&SimpleParameter::paramChangeCallback, this, _1));
    }

private:
    OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;

    rcl_interfaces::msg::SetParametersResult paramChangeCallback(const std::vector<rclcpp::Parameter> & parameters) {
        rcl_interfaces::msg::SetParametersResult result;

        for(const auto& param : parameters) {
            if(param.get_name() == "simple_int_param" && param.get_type() == rclcpp::ParameterType::PARAMETER_INTEGER) {
                RCLCPP_INFO_STREAM(get_logger(), "New param: " << param.as_int());
                result.successful = true;
            }

            if(param.get_name() == "simple_string_param" && param.get_type() == rclcpp::ParameterType::PARAMETER_STRING) {
                RCLCPP_INFO_STREAM(get_logger(), "New param: " << param.as_string()  << ", Old Param: " << get_parameter("simple_string_param").as_string());
                result.successful = true;
            }
        }

        return result;
    }
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleParameter>();
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}